torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
-  ys = tree_map(lambda *x: torch.stack(x), *ys)
+  y_len = len(fake_output_y)
+  fn_outputs = fake_output_carry + fake_output_y
+
+  fn_ctx = torch_xla._XLAC.lowering.LoweringContext()
+  fn_ctx.set_name_string("fn_ctx")
+  fn_ctx.build(list(fn_outputs))
+  fn_hlo = fn_ctx.hlo()
+  fn_computation = xb.computation_from_module_proto("fn_computation", fn_hlo)
+
+  # Figure out the shape of `ys` from the abstract tracing.
+  fn_carry_out, fn_y_out = split(fn_outputs, carry_len)
+  assert carry_len + y_len == len(fn_outputs)
+  fn_carry_shapes = [v.shape for v in fn_carry_out]
+  fn_y_shapes = [v.shape for v in fn_y_out]
+  for fn_carry_shape, init_leaf in zip(fn_carry_shapes, init):
+    assert fn_carry_shape == init_leaf.shape, f"`fn` must keep the `carry` shape unchanged. \
+      Got {fn_carry_shape} but expected {init_leaf.shape}"
+
+  builder = Builder('scan')
+  num_iters = next(iter(tree_iter(xs))).size(0)
+  ys = [
+      torch.zeros((num_iters, *fn_y_shape), device=device)
+      for fn_y_shape in fn_y_shapes
+  ]
+  # Start the `curr_iter` loop variable at zero.
+  zero = torch.tensor(0, device=device)
+  builder.add_param(zero)
+
+  # We are building a bigger XLA computation (the while loop) that calls
+  # a smaller computation (`fn_computation`). This is a mapping from
+  # `fn_computation` param ID to While computation param ID.
+  fn_param_id_to_while_param_id: Dict[int, int] = {}
+
+  # Add carry and x.
+  for real, fake in ((init, fake_carry), (xs, fake_x)):
+    for val, fake_val in zip(real, fake):
+      idx = builder.add_param(val)
+      param_id = fn_ctx.tensor_parameter_id(fake_val)
+      if param_id != -1:
+        fn_param_id_to_while_param_id[param_id] = idx
+
+  # Add the output as a param since our While computation consumes it, updates
+  # one slice, and returns the updated ys in each iteration.
+  for val in ys:
+    builder.add_param(val)
+
+  # Detect hoisted variables.
+  hoisted_vars: Dict[int, torch.Tensor] = fn_ctx.parameter_id_tensor_mapping()
+  for v in itertools.chain(fake_carry, fake_x):
+    param_id = fn_ctx.tensor_parameter_id(v)
+    if param_id != -1:
+      del hoisted_vars[param_id]
+
+  # Detect RNG seed usage within the scanned function within hoisted variables.
+  ids, i_values = torch_xla._XLAC._get_tensors_xla_device_data_node(fn_outputs)
+  seed_info_id = torch_xla._XLAC._get_seed_info_id()
+  seed_parameter_id = None
+  if seed_info_id in ids:
+    seed_idx = ids.index(seed_info_id)
+    seed_parameter_id = fn_ctx.tensor_parameter_id(i_values[seed_idx])
+    assert seed_parameter_id != -1, "`fn` uses random seed, but random seed is not \
+      a parameter to the traced HLO graph"
+
+    # Replace the single seed value with a tensor of seeds, one per iteration.
+    seed_tensor = hoisted_vars[seed_parameter_id]
+    assert seed_tensor.dtype == torch.int64
+    hoisted_vars[seed_parameter_id] = torch.randint(
+        0, 2**62, (num_iters,), dtype=torch.int64, device=torch_xla.device())
```

Comment:
in the future we should have a way to let user control seeds they want to use. It seems like right now if they do 
```
torch.manual_seed(xxx)
```
it will kind of work too as that will control the output of the `torch.randint` but this is buried too deep for most users...

Reply:
*nod* added a TODO